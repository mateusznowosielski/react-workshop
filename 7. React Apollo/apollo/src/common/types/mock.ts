export declare interface QueryMock {
  request: {
    query?: object,
    variables?: object,
  },
  result?: {
    data?: object | undefined,
    errors?: any,
  }
}