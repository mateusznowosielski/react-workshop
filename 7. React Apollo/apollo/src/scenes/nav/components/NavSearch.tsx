import * as React from 'react';
import { useState } from 'react';
import {Button, Input} from 'antd';
import {QueryHookResult, useQuery} from "@dc0de/react-apollo-hooks";
import {SEARCH_PCP} from "scenes/nav/components/queries/rest";
import useRouterHook from 'use-react-router';

const NavSearch: React.FunctionComponent<any> = ()  => {

    const { history }: any = useRouterHook();
    const [ value, setValue ] = useState('');

    const {data, error, loading}: QueryHookResult<any, any> = useQuery(SEARCH_PCP, {
        suspend: false,
        skip: !value,
        variables: {
            key: value,
            limit: 1,
        }
    });

    const provider: any = data && data.searchPCP && data.searchPCP.data ? data.searchPCP.data[0] : null;

    return (
        <div>
            <Input
              placeholder="Navigate to..."
              onChange={(event: React.ChangeEvent<HTMLInputElement>) => setValue(event.target.value)}
            />
            {
                loading && <p>Loading...</p>
            }
            {
                error && <p>{error.message}</p>
            }
            {
                provider && <Button onClick={() => { history.push(`/provider/${provider.id}`) }}>Go</Button>
            }
        </div>
    );
};

export default NavSearch;