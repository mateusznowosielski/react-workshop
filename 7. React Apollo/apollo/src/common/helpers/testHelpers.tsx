import * as React from 'react';
import { ApolloClient } from 'apollo-client';
import { InMemoryCache } from 'apollo-cache-inmemory';
import { MockLink } from 'apollo-link-mock';
import { Route, Router } from 'react-router-dom';
import {createMemoryHistory} from 'history';
import {render} from 'react-testing-library';
import { ApolloProvider } from 'react-apollo';
import { ApolloProvider as ApolloHooksProvider } from '@dc0de/react-apollo-hooks';
import { QueryMock } from '../types/mock';

export const MOCK: Array<QueryMock> = [
  {
    request: {},
    result: {},
  }
];

export function renderWithRouter (
  ui: React.ReactElement<any>,
  {route = '/', history = createMemoryHistory({initialEntries: [route]})} = {},
) {
  return {
    ...render(<Router history={history}>{ui}</Router>),
    history,
  }
}

// export const routerMock = () => ({
//   push: jest.fn(),
//   replace: jest.fn(),
//   go: jest.fn(),
//   goBack: jest.fn(),
//   goForward: jest.fn(),
//   setRouteLeaveHook: jest.fn(),
//   isActive: jest.fn(),
// });

export function createClient(mocks: any = MOCK) {
  return new ApolloClient({
    cache: new InMemoryCache({addTypename: false}),
    link: new MockLink(mocks),
  });
}

// export function renderWithApolloHooks(
//   ui: React.ReactElement<any>,
//   client: ApolloClient<any> = createClient()
// ) {
//   return {
//     ...render(
//       <ApolloProvider client={client}>
//         <ApolloHooksProvider client={client}>
//           {ui}
//         </ApolloHooksProvider>
//       </ApolloProvider>
//     ),
//     client,
//   }
// }

export function renderWithApolloHooksRouter(
  ui: React.ReactElement<any>,
  {
    client = createClient(),
    route = '/',
    history = createMemoryHistory({initialEntries: [route]})
  } = {},
) {
  return {
    ...render(
      <ApolloProvider client={client}>
        <ApolloHooksProvider client={client}>
          <Router history={history}>
            {ui}
          </Router>
        </ApolloHooksProvider>
      </ApolloProvider>
    ),
    client,
    history,
  }
}

export function renderWithApolloHooksRouterRoute(
  ui: React.ComponentType<any>,
  {
    client = createClient(),
    route = '/',
    path = '',
    history = createMemoryHistory({initialEntries: [route]})
  } = {},
) {
  return {
    ...render(
      <ApolloProvider client={client}>
        <ApolloHooksProvider client={client}>
          <Router history={history}>
            <Route component={ui} path={path} />
          </Router>
        </ApolloHooksProvider>
      </ApolloProvider>
    ),
    client,
    history,
  }
}

export const waitForNextTick = () => new Promise(resolve => setTimeout(resolve));

