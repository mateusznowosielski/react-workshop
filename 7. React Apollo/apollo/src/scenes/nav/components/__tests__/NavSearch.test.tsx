import * as React from 'react';
import { cleanup, fireEvent, render } from 'react-testing-library';
import NavSearch from 'scenes/nav/components/NavSearch';
import { createClient, renderWithApolloHooksRouter, waitForNextTick } from 'common/helpers/testHelpers';
import { MOCK_ERROR, MOCK_SUCCESS } from 'scenes/nav/components/__mocks__/navSearch.mock';
import ApolloClient from 'apollo-client';

afterEach(cleanup);

it('verifying NavSearch to be rendered', () => {
  const { container } = renderWithApolloHooksRouter(<NavSearch/>);

  expect(container).toMatchSnapshot();
});

it('verifying NavSearch to be rendered with GO button', async () => {
  const client: ApolloClient<any> = createClient(MOCK_SUCCESS);
  const { container } = renderWithApolloHooksRouter(<NavSearch />, { client });

  const input: HTMLInputElement = container.querySelector(`input`) as HTMLInputElement;
  fireEvent.click(input);

  // @ts-ignore
  render(null);
  await waitForNextTick();

  expect(container).toMatchSnapshot();
});

it('verifying NavSearch to be rendered with error', async () => {
  const client: ApolloClient<any> = createClient(MOCK_ERROR);
  const wrapper = renderWithApolloHooksRouter(<NavSearch />, { client });

  const input: HTMLInputElement = wrapper.container.querySelector('input') as HTMLInputElement;
  fireEvent.click(input);
  fireEvent.change(input, { target: { value: 'None' }});

  // @ts-ignore
  render(null);
  await waitForNextTick();

  expect(wrapper.getAllByText('Error!')).toBeDefined();
});