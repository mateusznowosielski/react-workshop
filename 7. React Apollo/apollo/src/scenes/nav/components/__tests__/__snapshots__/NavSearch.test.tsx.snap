// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`verifying NavSearch to be rendered 1`] = `
<div>
  <div>
    <input
      class="ant-input"
      placeholder="Navigate to..."
      type="text"
    />
  </div>
</div>
`;

exports[`verifying NavSearch to be rendered in loading state 1`] = `
<div>
  <div>
    <input
      class="ant-input"
      placeholder="Navigate to..."
      type="text"
    />
    <p>
      Loading...
    </p>
  </div>
</div>
`;
